File: LeetCode/215.cpp
//
//  main.cpp
//  LeetCode
//
//  Created by Hao on 2017/3/16.
//  Copyright © 2017年 Hao. All rights reserved.
//

#include <iostream>
#include <vector>
#include <queue>
#include <set>
using namespace std;

class Solution {
public:
    // STL sort
    int findKthLargest(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end()); // STL sort in ascending order
        return nums.at(nums.size() - k);
    }
    
    // Divide and Conquer
    int findKthLargest2(vector<int>& nums, int k) {
        int l = 0, r = nums.size() - 1;
        
        while (true) {
            int pos = partition(nums, l, r); // one time quick sort
            
            if (pos == k - 1) // the k-th largest key
                return nums.at(k - 1);
            else if (pos < k - 1) // in the right partition of pos
                l = pos + 1;
            else // in the left partition of pos
                r = pos - 1;
        }
    }
    
    int partition(vector<int> & nums, int left, int right) {
        int i = left, j = right, pivot = nums.at(left); // pivot key
        
        while (i < j) {
            while (i < j && nums.at(j) <= pivot) -- j; // scan from right to left
            if (i < j) nums.at(i ++) = nums.at(j); // place nums[j]
            
            while (i < j && nums.at(i) >= pivot) ++ i; // scan from left to right
            if (i < j) nums.at(j --) = nums.at(i); // place nums[i]
        }
        
        nums.at(i) = pivot; // place pivot key - the (i+1)-th largest key
        
        return i;
    }
    
    // priority_queue
    int findKthLargest3(vector<int>& nums, int k) {
        priority_queue<int> pq(nums.begin(), nums.end());
        
        for (int i = 0; i < k - 1; i ++)
            pq.pop(); // remove k-1 largest elements
        
        return pq.top(); // Kth largest element
    }

    // multiset
    int findKthLargest4(vector<int>& nums, int k) {
        multiset<int>   mset;
        
        for (int i = 0; i < nums.size(); ++ i) {
            mset.insert(nums.at(i));
            
            if (mset.size() > k)
                mset.erase(mset.begin()); // erase the element < Kth largest element
        }
        
        return *mset.begin(); // Kth largest element
    }
};

int main(int argc, char* argv[])
{
    Solution    testSolution;
    
    vector<vector<int>>     aInputs = {{3,2,1,5,6,4}, {1,2,3,4,5}, {10,9,8,7,6}};
    vector<int>             kInputs = {2, 4, 2};
    int                     result;
    
    /*
     5
     2
     9
     */
    for (auto i = 0; i < aInputs.size(); ++ i) {
        result = testSolution.findKthLargest(aInputs[i], kInputs[i]);
        cout << result << endl;
        result = testSolution.findKthLargest2(aInputs[i], kInputs[i]);
        cout << result << endl;
        result = testSolution.findKthLargest3(aInputs[i], kInputs[i]);
        cout << result << endl;
        result = testSolution.findKthLargest4(aInputs[i], kInputs[i]);
        cout << result << endl;
    }
    
    return 0;
}

File: 学习笔记之LeetCode/215.py
class Solution:
    # sorted()
    def findKthLargest(self, nums: List[int], k: int) -> int:
        return sorted(nums, reverse=True)[k - 1]
            
    # heapq.nlargest()
    def findKthLargest2(self, nums: List[int], k: int) -> int:
        return heapq.nlargest(k, nums)[-1]

    # Max Heap with heapq
    def findKthLargest3(self, nums: List[int], k: int) -> int:
        nums = [-num for num in nums]
        
        heapq.heapify(nums)
        
        for _ in range(k):
            result = heapq.heappop(nums)
            
        return -result    
    
    # Min Heap with heapq
    def findKthLargest4(self, nums: List[int], k: int) -> int:
        min_heap = [-float('inf')] * k
        
        # min heap with k elements
        heapq.heapify(min_heap)
        
        for num in nums:
            if num > min_heap[0]:
                heapq.heappop(min_heap)
                heapq.heappush(min_heap, num)
        
        return min_heap[0]
    
    # Divide and Conquer
    def findKthLargest5(self, nums: List[int], k: int) -> int:
        _left, _right = 0, len(nums) - 1
        
        def partition(nums, left, right):
            i, j, pivot = left, right, nums[left]
            
            while i < j:
                while i < j and nums[j] <= pivot: 
                    j -= 1
                if i < j:
                    nums[i] = nums[j]
                    i += 1
                
                while i < j and nums[i] >= pivot:
                    i += 1
                if i < j:
                    nums[j] = nums[i]
                    j -= 1
            
            nums[i] = pivot
            
            return i
        
        while True:
            pos = partition(nums, _left, _right)
            
            if pos == k - 1:
                return nums[k - 1]
            elif pos < k - 1:
                _left = pos + 1
            else:
                _right = pos - 1
